### DISCUSSION TITLE
TypeGuard for NamedTuple?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#5116] [Q&A] [answered]
I am trying to improve the type hints in PyTorch's `torch.nn.parallel` module. [There is a function they are using that is essentially a `TypeGuard` for `NamedTuple`](https://github.com/pytorch/pytorch/blob/ee40cce475705a9e47321a7db179102b50cfc12f/torch/nn/parallel/scatter_gather.py#L12). When I make it so, and based on how the narrowed value is being used, something interesting happens:

```python
from typing import Any, NamedTuple, TypeGuard


def _is_namedtuple(obj: Any) -> TypeGuard[NamedTuple]:
    # Check if type was created from collections.namedtuple or a typing.NamedTuple.
    return (
        isinstance(obj, tuple) and hasattr(obj, "_asdict") and hasattr(obj, "_fields")
    )


def scatter_map(obj):
    ...
    if _is_namedtuple(obj):
        return [type(obj)(*args) for args in zip(*map(scatter_map, obj))]  # error
    ...
    return [obj for _ in []]

```

```
Expected named tuple entry list as second argument
Expected named tuple class name as first argument
```

It seems like the constructor for the [typeshed definition of the `TypeGuard` class](https://github.com/python/typeshed/blob/2d5dafadb75f122e0c2aeced09a5b72232373f40/stdlib/typing_extensions.pyi#L252) is what is getting invoked here for `type(obj)`. This seems like a unique situation because logical instances of this class don't actually have this constructor... they accept `*args, **kwargs` based on the fields their tuple subclass contains.

Do you have any advice for how to type this? Do you see any improvements to how I have typed this or any nuance that could be added to the typeshed definition?

**Comment by erictraut** [comment, marked as answer]
`NamedTuple` isn't a type in the Python type system. It's a special form that allows you to define named tuple classes. The lowercase form (`namedtuple`) is a function that acts as a class factory. At runtime, the type of a named tuple is simply `tuple`.

```python
x = namedtuple("x", "a b c")
print(x.__mro__) # (x, tuple, object)
```

With this in mind, using `NamedTuple` in a type annotation doesn't make sense. It is not considered a subtype or supertype of any other type, so it doesn't have any meaning in type narrowing contexts. You're likely to get nonsensical errors if you attempt to use it in that manner.

**Reply by erictraut** [reply]
Depending on your objective, you might be able to accomplish it by defining your own protocol class.
